torch2.7 训练 开启cppwrapper报错

### Describe the bug🐛

哦哦 刚刚误解了，报错的信息我没截图，有记录
Traceback (most recent call last):
  File "/data01/tanhaojue/torch_train/0630_qianchuan_1B/modelcode/main_model.py", line 2763, in <module>
    runner.run()
  File "/usr/local/lib/python3.11/site-packages/bytedance/lagrange_torch/runner/cmd_base_runner.py", line 31, in run
    return self.run_internal(**kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data01/tanhaojue/torch_train/0630_qianchuan_1B/modelcode/main_model.py", line 2704, in run_internal
    result = runstep.func(**runstep.module_dict)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data01/tanhaojue/torch_train/0630_qianchuan_1B/modelcode/main_model.py", line 2615, in oracle_train
    result, loss = root_module()
                   ^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1751, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1762, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/torch_mlu/_dynamo/eval_frame.py", line 189, in _fn
    raise e.remove_dynamo_frames() from None  # see TORCHDYNAMO_VERBOSE=1
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/torch/_inductor/compile_fx.py", line 760, in _compile_fx_inner
    raise InductorError(e, currentframe()).with_traceback(
  File "/usr/local/lib/python3.11/site-packages/torch/_inductor/compile_fx.py", line 745, in _compile_fx_inner
    mb_compiled_graph = fx_codegen_and_compile(
                        ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/torch/_inductor/compile_fx.py", line 1295, in fx_codegen_and_compile
    return scheme.codegen_and_compile(gm, example_inputs, inputs_to_check, graph_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/torch/_inductor/compile_fx.py", line 1197, in codegen_and_compile
    compiled_fn = graph.compile_to_module().call
                  ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/torch/_inductor/graph.py", line 2083, in compile_to_module
    return self._compile_to_module()
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/torch/_inductor/graph.py", line 2091, in _compile_to_module
    self.codegen_with_cpp_wrapper() if self.cpp_wrapper else self.codegen()
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/torch/_inductor/graph.py", line 1981, in codegen_with_cpp_wrapper
    return self.codegen()
           ^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/torch/_inductor/graph.py", line 2035, in codegen
    result = self.wrapper_code.generate(self.is_inference)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/torch_mlu/_inductor/codegen/cpp_wrapper_mlu.py", line 85, in generate
    return super().generate(is_inference)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/torch/_inductor/codegen/cpp_wrapper_gpu.py", line 303, in generate
    return super().generate(is_inference)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/torch/_inductor/codegen/cpp_wrapper_cpu.py", line 883, in generate
    return super().generate(is_inference)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/torch/_inductor/codegen/wrapper.py", line 1197, in generate
    return self._generate(is_inference)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/torch/_inductor/codegen/wrapper.py", line 1263, in _generate
    self.finalize_prefix()
  File "/usr/local/lib/python3.11/site-packages/torch/_inductor/codegen/cpp_wrapper_gpu.py", line 312, in finalize_prefix
    kernel.generate(self)
  File "/usr/local/lib/python3.11/site-packages/torch/_inductor/codegen/cpp_wrapper_gpu.py", line 65, in generate
    assert params, f"CudaKernelParamCache not populated for {self.kernel_name}"
           ^^^^^^
torch._inductor.exc.InductorError: AssertionError: CudaKernelParamCache not populated for triton_poi_fused_0

Set TORCHDYNAMO_VERBOSE=1 for the internal stack trace (please do this especially if you're reporting a bug to PyTorch). For even more developer context, set TORCH_LOGS="+dynamo"

### Environment details

PyTorch:2.7

--- a/.github/workflows/ci-for-mlu.yml
+++ b/.github/workflows/ci-for-mlu.yml
@@ -48,7 +48,7 @@ jobs:
     name: Run mlu tests
     runs-on: MLU
     container:
-      image: hub.byted.org/ci/xpu_graph:934ebcfdd670bcea94aa0a0ed247a576
+      image: hub.byted.org/ci/xpu_graph:c5672b8705309273fc76f33d6eeefcfd
       options: --privileged=true --ipc=host --pid=host --user root --network host
     steps:
     - uses: actions/checkout@v4

--- a/tests/mlu/test_cppwrapper.py
+++ b/tests/mlu/test_cppwrapper.py
@@ -1,0 +1,61 @@
+import pytest
+import torch
+import torch.nn.functional as F
+import torch_mlu
+
+import xpu_graph
+from xpu_graph.config import OptLevel
+from xpu_graph.test_utils import (
+    assertTensorsEqual,
+    need_xpu_graph_logs,
+    skip_xpu_graph_cache,
+)
+
+device = "mlu:0"
+data_type = torch.float32
+aten = torch.ops.aten
+
+
+def fn0(input1, input2):
+    return input1 + input2
+
+
+def matmul_test(xpu_graph_backend, func):
+    input1 = torch.randn((4096, 768), dtype=data_type).to(device)
+    input2 = torch.randn((4096, 768), dtype=data_type).to(device)
+    compiled = torch.compile(func, backend=xpu_graph_backend, dynamic=False)
+    res1 = compiled(input1, input2)
+    res = func(input1, input2)
+    assertTensorsEqual(
+        res.cpu().float(), res1.cpu().float(), 0.005, use_MSE=True, use_RAE=True
+    )
+
+
+class TestMatMul:
+    def setup_class(self):
+        self.xpu_graph_backend = xpu_graph.mlu_compiler(
+            is_training=False,
+            opt_level=OptLevel.level2,
+            vendor_compiler_config={"mode": "default", "cpp_wrapper": True},
+        )
+
+    @pytest.mark.parametrize(
+        "pattern_func",
+        [
+            fn0,
+        ],
+    )
+    def test_matmul_patterns(self, caplog, pattern_func):
+        with need_xpu_graph_logs(), skip_xpu_graph_cache(self.xpu_graph_backend):
+            matmul_test(self.xpu_graph_backend, pattern_func)
+        if pattern_func in [fn0]:
+            assert "Pattern.FusedMatMul changed graph" in caplog.text
+
+
+if __name__ == "__main__":
+    xpu_compiler_backend = xpu_graph.mlu_compiler(
+        is_training=False,
+        opt_level=OptLevel.level2,
+        vendor_compiler_config={"mode": "default", "cpp_wrapper": True},
+    )
+    matmul_test(xpu_compiler_backend, fn0)

--- a/tests/mlu/test_cppwrapper.py
+++ b/tests/mlu/test_cppwrapper.py
@@ -17,6 +17,10 @@ aten = torch.ops.aten
 
 
 def fn0(input1, input2):
+    return input2 @ input1.T
+
+
+def fn1(input1, input2):
     return input1 + input2
 
 
@@ -39,9 +43,7 @@ class TestMatMul:
 
     @pytest.mark.parametrize(
         "pattern_func",
-        [
-            fn0,
-        ],
+        [fn0, fn1],
     )
     def test_matmul_patterns(self, caplog, pattern_func):
         with need_xpu_graph_logs(), skip_xpu_graph_cache(self.xpu_graph_backend):

--- a/tests/mlu/test_cppwrapper.py
+++ b/tests/mlu/test_cppwrapper.py
@@ -26,9 +26,7 @@ def matmul_test(xpu_graph_backend, func):
     compiled = torch.compile(func, backend=xpu_graph_backend, dynamic=False)
     res1 = compiled(input1, input2)
     res = func(input1, input2)
-    assertTensorsEqual(
-        res.cpu().float(), res1.cpu().float(), 0.005, use_MSE=True, use_RAE=True
-    )
+    assertTensorsEqual(res.cpu().float(), res1.cpu().float(), 0.005, use_MSE=True, use_RAE=True)
 
 
 class TestMatMul:

--- a/xpu_graph/backends/mlu.py
+++ b/xpu_graph/backends/mlu.py
@@ -14,11 +14,7 @@ def mlu_compile(module: torch.nn.Module, example_inputs, config_dict: Dict, **kw
         return cudagraphs(module, example_inputs)
 
     from torch import _TorchCompileInductorWrapper
-    from torch._inductor.compile_fx import (
-        compile_fx,
-        compile_fx_inner,
-        get_cpp_wrapper_config,
-    )
+    from torch._inductor.compile_fx import compile_fx, compile_fx_inner
 
     options = {}
     dynamic = config_dict.get("dynamic", True)
@@ -27,7 +23,12 @@ def mlu_compile(module: torch.nn.Module, example_inputs, config_dict: Dict, **kw
         from packaging import version
 
         torch_version = version.parse(torch.__version__[:5])
-        config = get_cpp_wrapper_config() if cpp_wrapper and torch_version >= version.parse("2.7.0") else {}
+        if cpp_wrapper and torch_version >= version.parse("2.7.0"):
+            from torch._inductor import get_cpp_wrapper_config
+
+            config = get_cpp_wrapper_config()
+        else:
+            config = {}
         with torch._inductor.config.patch(**config):
             compiled_func = compile_fx_inner(module, example_inputs, cpp_wrapper=cpp_wrapper, **kwargs)
     return compiled_func

--- a/xpu_graph/backends/mlu.py
+++ b/xpu_graph/backends/mlu.py
@@ -14,12 +14,20 @@ def mlu_compile(module: torch.nn.Module, example_inputs, config_dict: Dict, **kw
         return cudagraphs(module, example_inputs)
 
     from torch import _TorchCompileInductorWrapper
-    from torch._inductor.compile_fx import compile_fx, compile_fx_inner
+    from torch._inductor.compile_fx import (
+        compile_fx,
+        compile_fx_inner,
+        get_cpp_wrapper_config,
+    )
 
     options = {}
     dynamic = config_dict.get("dynamic", True)
     inductor_backend = _TorchCompileInductorWrapper(mode, options, dynamic)
     with torch._inductor.config.patch(inductor_backend.config):
-        compiled_func = compile_fx_inner(module, example_inputs, cpp_wrapper=cpp_wrapper, **kwargs)
+        from packaging import version
 
+        torch_version = version.parse(torch.__version__[:5])
+        config = get_cpp_wrapper_config() if cpp_wrapper and torch_version >= version.parse("2.7.0") else {}
+        with torch._inductor.config.patch(**config):
+            compiled_func = compile_fx_inner(module, example_inputs, cpp_wrapper=cpp_wrapper, **kwargs)
     return compiled_func

--- a/xpu_graph/passes/patterns/targets/mlu/combo_slice_where_cat.py
+++ b/xpu_graph/passes/patterns/targets/mlu/combo_slice_where_cat.py
@@ -63,10 +63,12 @@ class ComboSliceWhereCat(Pattern):
             if node.target == torch.ops.aten.stack.default:
                 is_stack = True
             ori_cat_input = node.args[0]
-            if is_stack:
-                axis = 0
-            else:
+            if "dim" in node.kwargs:
+                axis = node.kwargs["dim"]
+            elif len(node.args) > 1:
                 axis = node.args[1]
+            else:
+                axis = 0
             if len(ori_cat_input) < MINI_LEN:
                 continue
             start, end = match_sub_list(

